index.bs

```diff
@@ -9,8 +9,187 @@ Editor: Kent Tamura, Google https://www.google.com/, tkent@google.com
+    this.addEventListener('click', this._onClick.bind(this));
+
+    const shadowRoot = this.attachShadow({mode: 'closed'});
+    shadowRoot.innerHTML =
+      &#96;&lt;style>
+       :host::before {
+         content: '[ ]';
+         white-space: pre;
+         font-family: monospace;
+       }
+       :host(:state(checked))::before { content: '[x]' }
+       &lt;/style>
+       &lt;slot>Label&lt;/slot>&#96;;
+  }
+
+  get checked() { return this._internals.states.contains('checked'); }
+
+  set checked(flag) {
+    this._internals.states.toggle('checked', !!flag);
+  }
+
+  _onClick(event) {
+    this.checked = !this.checked;
+  }
+}
+
+customElements.define('labeled-checkbox', LabeledCheckbox);
+&lt;/script>
+
+&lt;style>
+labeled-checkbox { border: dashed red; }
+labeled-checkbox:state(checked) { border: solid; }
+&lt;/style>
+
+&lt;labeled-checkbox>You need to check this&lt;/labeled-checkbox>
+
+<!-- Works even on ::part()s -->
+&lt;script>
+class QuestionBox extends HTMLElement {
+  constructor() {
+    super();
+    const shadowRoot = this.attachShadow({mode: 'closed'});
+    shadowRoot.innerHTML =
+      &#96;&lt;div>&lt;slot>Question&lt;/slot>&lt;/div>
+       &lt;labeled-checkbox part='checkbox'>Yes&lt;/labeled-checkbox>&#96;;
+  }
+}
+customElements.define('question-box', QuestionBox);
+&lt;/script>
+
+&lt;style>
+question-box::part(checkbox) { color: red; }
+question-box::part(checkbox):state(checked) { color: green; }
+&lt;/style>
+
+&lt;question-box>Continue?&lt;/question-box>
+&lt;/body>
+</pre>
+</div>
+
+
+Exposing custom element states {#exposing}
+============================
+
+Each <a>autonomous custom element</a> has <dfn>states token list</dfn>, a
+<a>non-attribute <code>DOMTokenList</code></a> object associated with the custom
+element, and initially associated with an empty [=/set=] of strings.
+
+<pre class=idl>
+partial interface ElementInternals {
+  [SameObject, PutForwards=value] readonly attribute DOMTokenList states;
+};
+</pre>
+
+The {{states}} IDL attribute returns the [=states token list=] of 
+this's
+<a href="http://html.spec.whatwg.org/C/#internals-target">target element</a>.
+
+<div class="example" id="ex-non-boolean-state">
+[=States token list=] can expose boolean states represented by
+existence/non-existence of string tokens. If an author wants to expose a state
+which can have three values, it can be converted to three exclusive boolean
+states. For example, a state called <code>readyState</code> with
+<code>"loading"</code>, <code>"interactive"</code>, and <code>"complete"</code>
+values can be mapped to three exclusive boolean states, <code>"loading"</code>,
+<code>"interactive"</code>, and <code>"complete"</code>.
+
+<pre class="lang-js">
+// Change the readyState from anything to "complete".
+this._readyState = "complete";
+this.internals.<l>{{states}}</l>.<l>{{DOMTokenList/remove}}</l>("loading", "interactive");
+this.internals.<l>{{states}}</l>.<l>{{DOMTokenList/add}}</l>("complete");
```

Comment:
I just realized you could also write this as

```js
this.internals.states = "complete";
```

due to the PutForwards. Maybe it's worth including that in addition to what you have.

Reply:
It's a good idea.  Done.